src/core/regimes.rkt

```diff
@@ -22,19 +22,54 @@
 ;; pidx = Point index: The index of the point to the left of which we should split.
 (struct si (cidx pidx) #:prefab)
 
+(define (pareto-regimes sorted ctx)
+  (define branch-exprs
+    (if (flag-set? 'reduce 'branch-expressions)
+        (exprs-to-branch-on sorted ctx)
+        (context-vars ctx)))
+  (define err-lsts (batch-errors (map alt-expr sorted) (*pcontext*) ctx))
+  (define init-errs (for/hash ([bexpr branch-exprs]) (values bexpr -1)))
+  (let loop ([alts sorted] [errs init-errs])
+    (cond
+     [(or (null? alts) (= (length alts) 1)) '()]
+     [else
+      (define recomputed-branch-exprs
+        (if (flag-set? 'reduce 'branch-expressions)
+            (exprs-to-branch-on alts ctx)
+            (context-vars ctx)))
+      
+      (define recomp-errs 
+        (for/hash ([bexpr recomputed-branch-exprs]) (values bexpr (hash-ref errs bexpr))))
+
+      (define-values (opt new-errs) 
+        (infer-splitpoints alts recomputed-branch-exprs recomp-errs ctx))
+      (define high (si-cidx (argmax (λ (x) (si-cidx x)) (option-split-indices opt))))
+      (cons opt (loop (take alts high) new-errs))])))
+
 ;; `infer-splitpoints` and `combine-alts` are split so the mainloop
 ;; can insert a timeline break between them.
 
-(define (infer-splitpoints alts ctx)
+(define (infer-splitpoints alts branch-exprs cerrs ctx)
   (timeline-event! 'regimes)
   (timeline-push! 'inputs (map (compose ~a alt-expr) alts))
-  (define branch-exprs
-    (if (flag-set? 'reduce 'branch-expressions)
-        (exprs-to-branch-on alts ctx)
-        (context-vars ctx)))
   (define err-lsts (batch-errors (map alt-expr alts) (*pcontext*) ctx))
-  (define options (for/list ([bexpr branch-exprs]) (option-on-expr alts err-lsts bexpr ctx)))
-  (define best (argmin (compose errors-score option-errors) options))
+  (define sorted-bexprs (sort branch-exprs (lambda (x y) (< (hash-ref cerrs x) (hash-ref cerrs y)))))
```

Comment:
You can actually pass `#:key (curry hash-ref cerrs)` to the sort function and then pass `<` for the comparator.